vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,330 @@
+    batch_size = size(x)[end]
+    # (batch) of random times to approximate 𝔼[⋅] wrt. 𝘪 ∼ 𝒰(0, 𝘛)
+    random_t = rand(Float32, batch_size) .* (1.0f0 - ϵ) .+ ϵ |> device
+    # (batch) of perturbations to approximate 𝔼[⋅] wrt. 𝘹(0) ∼ 𝒫₀(𝘹)
+    z = randn!(similar(x))
+    std = expand_dims(model.layers.marginal_prob_std(random_t), 3)
+    # (batch) of perturbed 𝘹(𝘵)'s to approximate 𝔼 wrt. 𝘹(t) ∼ 𝒫₀ₜ(𝘹(𝘵)|𝘹(0))
+    perturbed_x = x + z .* std
+    # 𝘚₀(𝘹(𝘵), 𝘵)
+    score = model(perturbed_x, random_t)
+    # mean over batches
+    sum(
+        # L₂ norm over WHC dimensions
+        sum(
+            (score .* std + z) .^ 2,
+            dims=(1, 2, 3)
+        )
+    ) / batch_size
+end
+
+"""
+Helper function that loads MNIST images and returns loader.
+"""
+function get_data(batch_size)
+    xtrain, ytrain = MLDatasets.MNIST.traindata(Float32)
+    xtrain = reshape(xtrain, 28, 28, 1, :)
+    DataLoader((xtrain, ytrain), batchsize=batch_size, shuffle=true)
+end
+
+"""
+Helper function from DrWatson.jl to convert a struct to a dict
+"""
+function struct2dict(::Type{DT}, s) where {DT<:AbstractDict}
+    DT(x => getfield(s, x) for x in fieldnames(typeof(s)))
+end
+struct2dict(s) = struct2dict(Dict, s)
+
+# arguments for the `train` function 
+@with_kw mutable struct Args
+    η = 1e-4                                        # learning rate
+    batch_size = 32                                 # batch size
+    epochs = 30                                     # number of epochs
+    seed = 1                                        # random seed
+    cuda = false                                    # use CPU
+    verbose_freq = 10                               # logging for every verbose_freq iterations
+    tblogger = true                                 # log training with tensorboard
+    save_path = "output"                            # results path
+end
+
+function train(; kws...)
+    # load hyperparamters
+    args = Args(; kws...)
+    args.seed > 0 && Random.seed!(args.seed)
+
+    # GPU config
+    if args.cuda && CUDA.has_cuda()
+        device = gpu
+        @info "Training on GPU"
+    else
+        device = cpu
+        @info "Training on CPU"
+    end
+
+    # load MNIST images
+    loader = get_data(args.batch_size)
+
+    # initialize UNet model
+    unet = UNet(marginal_prob_std) |> device
+
+    # ADAM optimizer
+    opt = ADAM(args.η)
+
+    # parameters
+    ps = Flux.params(unet)
+
+    !ispath(args.save_path) && mkpath(args.save_path)
+
+    # logging by TensorBoard.jl
+    if args.tblogger
+        tblogger = TBLogger(args.save_path, tb_overwrite)
+    end
+
+    # Training
+    train_steps = 0
+    @info "Start Training, total $(args.epochs) epochs"
+    for epoch = 1:args.epochs
+        @info "Epoch $(epoch)"
+        progress = Progress(length(loader))
+
+        for (x, _) in loader
+            loss, grad = Flux.withgradient(ps) do
+                model_loss(unet, x |> device, device)
```

Comment:
It's always better to move `x` to cpu/gpu _outside_ of the gradient callback unless absolutely necessary.

Reply:
Which also implies that `model_loss` shouldn't (have) to take a `device` argument either. Happy to elaborate once you've updated to use `similar` for most bits :)

Reply:
Ideally, the device on which some arrays live should already be defined before any function evaluation. That is to say that ideally we shouldn't need to have more than a couple invocations to `cpu/gpu` ahead of training.

Reply:
Would this be the fix?
```julia
      x = x |> device
      for (x, _) in loader
          loss, grad = Flux.withgradient(ps) do
              model_loss(unet, x, device)
```

Reply:
Did you mean:
```julia
for (x, _) in loader
  x = device(x)
  ...
```
?

Reply:
Oh, yep! I will add next commit

Reply:
I think the better solution is to do something like 

```julia

X = device(x)
model = device(x)

loader = DataLoader(X)
for in loader
...
```